### PR TITLE
ON HOLD - Issue 1489: Reduce email sends to first-time transition for the applicable transitions

### LINKED ISSUE #1489: Reduce email notifications for some status changes
### Issue description

As a domain requestor
I want to receive notifications of status changes ONLY when those changes initially occur
So that I do not get excessive emails which are not meaningful to me

Existing behavior: A notification email is sent to the requestor upon each occurrence of a status change. (Submitted, withdrawn, approved, rejected)

### Acceptance criteria

- [x] No notification is sent when the request moves from "Action Needed" to "Submitted" (requester re-submits with changes)
- [x] No notification is sent when the request moves from "In Review" back to "Submitted" (analyst returned it to the queue)
- [x] #1582

### Additional context

Consider the following statuses, listed horizontally, with the possible transitions listed vertically below each.
Image

For those statuses that have the corresponding email icon, a notification email should be sent when the request INITIALLY enters that status. Subsequent status changes may take place which might return the request to a previously held status. When this happens, a second email should NOT be sent to the requestor as this adds no valuable information and may even be confusing to the user.
 
For example, a request may be `submitted` by the requestor, which triggers a notification email confirming the submission. Later, the request may be moved to `In Review` (which does not use a notification). If this request is moved back to `submitted` status, this does NOT trigger another notification email, as it has already been in that status.

### Links to other issues

Related to:  #1582 

_No response_

### PR DESCRIPTION
**Pull request by rachidatecs**
…tions

## Ticket

Resolves #1489 

## Changes

- Query the logs to determine whether a request has ever been in a given previous state
- For the applicable FSM transitions, skip the email send if the above condition is true
- Unit tests on models and from the DjA perspective (different approaches that test the same thing)

## Context for reviewers

The model tests can be streamlined. I might revisit tomorrow.

## Setup

Create an application and go to town transitioning back an forth through the allowable transitions. You should only get the submitted, approved, withdrawn and rejected emails ONCE each, on first time transitioning to those states.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/domain_application.py [hunk 1]
```diff
@@ -633,11 +655,13 @@ def submit(self):
         self.submission_date = timezone.now().date()
         self.save()
 
-        self._send_status_update_email(
-            "submission confirmation",
-            "emails/submission_confirmation.txt",
-            "emails/submission_confirmation_subject.txt",
-        )
+        # Limit email notifications for this transition to the first time the request transitions to this status
+        if not self.has_previously_had_a_status_of("submitted"):
```
Comment (zandercymatics) [inline src/registrar/models/domain_application.py:659]:
```suggestion
        if not self.has_previously_had_a_status_of(DomainApplication.ApplicationStatus.SUBMITTED):
```
(Optional) May need some slight tweaking, but why not use the constant for this field and the others?

File: src/registrar/tests/test_admin.py [hunk 2]
```diff
@@ -421,29 +422,27 @@ def test_short_org_name_in_applications_list(self):
         # Now let's make sure the long description does not exist
         self.assertNotContains(response, "Federal: an agency of the U.S. government")
 
-    @boto3_mocking.patching
-    def test_save_model_sends_submitted_email(self):
-        # make sure there is no user with this email
-        EMAIL = "mayor@igorville.gov"
-        User.objects.filter(email=EMAIL).delete()
+    def transition_state_and_send_email(self, application, status):
```
Comment (zandercymatics) [inline src/registrar/tests/test_admin.py:425]: I'm 99% sure this is fine, but Boto is kind of temperamental, so I would just make sure that this method isn't firing out an email (you can add a log event to the `_send_status_update_email(` function I think)

File: src/registrar/tests/test_admin.py [hunk 3]
```diff
@@ -421,29 +422,27 @@ def test_short_org_name_in_applications_list(self):
...
                 request = self.factory.post("/admin/registrar/domainapplication/{}/change/".format(application.pk))
 
                 # Modify the application's property
-                application.status = DomainApplication.ApplicationStatus.SUBMITTED
+                application.status = status
 
                 # Use the model admin's save_model method
                 self.admin.save_model(request, application, form=None, change=True)
 
+    def assert_email_is_accurate(self, expected_string, email_index, email_address):
```
Comment (zandercymatics) [inline src/registrar/tests/test_admin.py:439]: Nice refactor!

**Review by dave-kennedy-ecs — APPROVED** [review]

File: src/registrar/tests/test_admin.py (hunk 2 above)
Comment (rachidatecs) [reply src/registrar/tests/test_admin.py:425]: Put a logger in there, didn't seem to trigger

File: src/registrar/models/domain_application.py [hunk 4]
```diff
@@ -578,7 +578,8 @@ def has_previously_had_a_status_of(self, status):
                 changes_dict = json.loads(entry.changes)
                 # changes_dict will look like {'status': ['withdrawn', 'submitted']},
                 # henceforth the len(changes_dict.get('status', [])) == 2
-                if len(changes_dict.get("status", [])) == 2 and changes_dict.get("status", [])[1] == status:
+                status_change = changes_dict.get("status", [])
+                if len(status_change) == 2 and status_change[1] == status:
```
Comment (zandercymatics) [inline src/registrar/models/domain_application.py:582]: Nice!

**Review by zandercymatics — APPROVED** [review]
LGTM

**Comment by rachidatecs** [comment]
The ACs changed and a new branch/PR have been made: https://github.com/cisagov/manage.get.gov/pull/1749

Look at this PR if we ever need to query or search audit logs.

**Pull request closed without merging**